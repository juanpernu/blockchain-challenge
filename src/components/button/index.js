import React from "react";

export const Button = ({ children, onClick }) => {
  return (
    <button
      className="rounded-md shadow mt-4 px-8 py-3 text-base font-medium rounded-md text-white bg-sky-500 hover:bg-sky-600"
      onClick={onClick}
    >
      {children}
    </button>
  );
};
