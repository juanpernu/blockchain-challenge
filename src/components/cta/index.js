import React from "react";

export const Cta = ({ children, className, onClick }) => {
  return (
    <button
      onClick={onClick}
      className={`bg-blue-500 hover:shadow-lg absolute bottom-4 right-4 mt-4 inline-flex h-14 w-14 w-full justify-center rounded-full p-2 text-xs font-light uppercase text-white shadow-sm transition ease-in-out hover:bg-blue-800 ${className}`}
    >
      {children}
    </button>
  );
};
