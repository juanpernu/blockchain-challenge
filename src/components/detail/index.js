import React from "react";
import { typesColors } from "../../utils";

export const Detail = ({ data }) => {
  const {
    name,
    id,
    types,
    height,
    weight,
    stats,
    sprites: { front_default },
  } = data;

  const getTypeClass = (type) => {
    if (typesColors[type]) return typesColors[type];
  };

  return (
    <div className="bg-white rounded-md p-8">
      <div className="flex flex-col">
        <p className="block text-sm font-medium text-gray-500 capitalize">
          # {id}
        </p>
        <p className="block text-2xl font-semibold text-gray-500 capitalize">
          {name}
        </p>
        <div className="flex mt-2">
          {types.map((t, i) => (
            <p
              className="capitalize text-sm mr-2 bg-slate-200 text-white px-2 rounded"
              style={{ backgroundColor: getTypeClass(t.type.name) }}
              key={i}
            >
              {t.type.name}
            </p>
          ))}
        </div>
      </div>
      <img
        src={front_default}
        width={300}
        height={300}
        className="m-auto"
        alt={`${name} front`}
      />
      <div className="flex justify-center">
        <p className="block text-sm font-medium text-gray-500 capitalize mr-8">
          Weight: {weight}kg
        </p>
        <p className="block text-sm font-medium text-gray-500 capitalize">
          Height: {height}m
        </p>
      </div>
      <div className="flex justify-between mt-8 px-8">
        {stats.map((s, i) => (
          <div className="flex flex-col text-center">
            <p className="m-0 p-0 text-sm font-bold">{s.base_stat}</p>
            <p className="m-0 p-0 text-xs">{s.stat.name}</p>
          </div>
        ))}
      </div>
    </div>
  );
};
