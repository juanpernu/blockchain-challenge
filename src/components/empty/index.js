import React from "react";

export const Empty = ({ message }) => (
  <p className="text-xl my-8 leading-8 font-extrabold text-center text-white">
    {message}
  </p>
);
