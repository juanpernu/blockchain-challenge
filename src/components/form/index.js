import React, { useState } from "react";
import { Button } from "../index";

export const Form = ({ onSubmit, onCancel }) => {
  const [newPokemon, setNewPokemon] = useState({});
  const [showError, setShowError] = useState({});

  const onChangeHandler = (e, property) => {
    let value = e.target.value;
    setNewPokemon({
      ...newPokemon,
      isNew: true,
      [property]: value,
    });
  };

  const onSubmitHandler = (e) => {
    e.preventDefault();
    if (!newPokemon.name && !newPokemon.id)
      return setShowError({ name: true, id: true });
    if (!newPokemon.name) return setShowError({ name: true, id: false });
    if (!newPokemon.id) return setShowError({ name: false, id: true });
    setShowError({});
    onSubmit(newPokemon);
  };

  const Error = () => (
    <p className="text-sm mt-2 text-red-400">* required field</p>
  );

  return (
    <form onSubmit={(e) => onSubmitHandler(e)}>
      <div className="grid mt-2 grid-cols-2 gap-x-8">
        <span className="relative">
          <label className="mb-2 block text-sm font-medium text-gray-700">
            ID *
          </label>
          <input
            type="number"
            id="id"
            onChange={(e) => onChangeHandler(e, "id")}
            className="rounded-md shadow-sm focus:ring-indigo-500 focus:border-indigo-500 block w-full pl-4 py-2 sm:text-sm border-gray-300 rounded-md"
          />
          {showError.id && <Error />}
        </span>
        <span className="relative">
          <label className="mb-2 block text-sm font-medium text-gray-700">
            Name *
          </label>
          <input
            type="text"
            id="name"
            onChange={(e) => onChangeHandler(e, "name")}
            className="rounded-md shadow-sm focus:ring-indigo-500 focus:border-indigo-500 block w-full pl-4 py-2 sm:text-sm border-gray-300 rounded-md"
          />
          {showError.name && <Error />}
        </span>
        <Button className="w-full" onClick={(e) => onSubmitHandler(e)}>
          <p>Add</p>
        </Button>
        <Button
          className="w-full bg-red-500 hover:bg-red-700"
          onClick={onCancel}
        >
          <p>Cancel</p>
        </Button>
      </div>
    </form>
  );
};
