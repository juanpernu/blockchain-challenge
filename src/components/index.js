export { Button } from "./button";
export { Detail } from "./detail";
export { Empty } from "./empty";
export { Item } from "./item";
export { Loading } from "./loading";
export { Pagination } from "./pagination";
