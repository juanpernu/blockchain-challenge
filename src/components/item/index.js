import React from "react";

export const Item = ({ data: { name, url, isNew = false }, onClick }) => {
  const [, id] = url.split("pokemon");
  const sanitazedId = id.replace("/", "").replace("/", "");
  return (
    <div
      className="flex justify-between border-b p-2 hover:bg-lime-200 hover:cursor-pointer"
      onClick={() => onClick(id)}
    >
      <div className="flex flex-col">
        <span className="block text-xs font-medium text-gray-500 capitalize">
          ID: {sanitazedId}
        </span>
        <p className="block text-base font-medium text-gray-700 mr-4 capitalize">
          {name}
        </p>
      </div>
      {isNew && (
        <span className="text-xs self-center rounded bg-sky-500 px-2 py-1 mr-4 font-semibold text-white shadow-md">
          NEW
        </span>
      )}
    </div>
  );
};
