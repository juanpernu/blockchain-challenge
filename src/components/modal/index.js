import React from "react";
import { XIcon } from "@heroicons/react/solid";
import { useKeypress } from "../../hooks";

export const Modal = ({ isOpenModal, closeModal, title, children }) => {
  useKeypress("Escape", closeModal);

  if (!isOpenModal) return null;

  return (
    <>
      <div className="overflow-y-auto overflow-x-hidden fixed bottom-1/2 z-20 w-auto h-modal rounded bg-white p-8 shadow-lg">
        <div className="flex flex-col">
          <span onClick={closeModal}>
            <XIcon
              className="mr-4 h-5 w-5 absolute top-6 right-4 text-slate-400 hover:text-slate-500 hover:cursor-pointer"
              aria-hidden="true"
            />
          </span>
          <h2 className="mt-4 text-xl font-semibold text-neutral-800">
            {title}
          </h2>
          {children}
        </div>
      </div>
      <div
        style={{ display: isOpenModal ? "block" : "none" }}
        className="fixed z-10 inset-0 bg-gray-600 bg-opacity-50 overflow-y-auto h-full w-full"
      />
    </>
  );
};
