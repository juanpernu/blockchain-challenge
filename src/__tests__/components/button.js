import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { Button } from "../../components";

const clickHandler = jest.fn();

test("render button and click on it", () => {
  const { rerender, asFragment } = render(
    <Button
      className="bg-sky-50 border-sky-500 text-sky-600 group"
      onClick={clickHandler}
    >
      <span className="sr-only">Anterior</span>
    </Button>
  );
  const button = screen.getByRole("button");

  expect(asFragment()).toMatchSnapshot();

  fireEvent.click(button);
  expect(clickHandler).toHaveBeenCalled();

  rerender(
    <Button
      className="bg-sky-50 border-sky-500 text-sky-600 group"
      onClick={clickHandler}
    >
      <span className="sr-only">Anterior</span>
    </Button>
  );
});
