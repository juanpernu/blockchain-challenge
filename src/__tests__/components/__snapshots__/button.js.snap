// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render button and click on it 1`] = `
<DocumentFragment>
  <button
    class="rounded-md shadow mt-4 px-8 py-3 text-base font-medium rounded-md text-white bg-sky-500 hover:bg-sky-600 bg-sky-50 border-sky-500 text-sky-600 group"
  >
    <span
      class="sr-only"
    >
      Anterior
    </span>
  </button>
</DocumentFragment>
`;
